#143 Product Page: Who Should Enroll

ahmed-arbisoft opened the pull request:
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested


#### What are the relevant tickets?
Closes #143 

#### What's this PR do?
Adds "Who Should Enroll" section to Course Page.

#### How should this be manually tested?
Create a "Who should enroll page" type child page on a `CoursePage`, visit the course page and the section should be visible.

#### Where should the reviewer start?
Assuming there are courses in the database and CMS, create a child page as detailed above.

#### Screenshots (if appropriate)
![desktop-who-should-enroll](https://user-images.githubusercontent.com/45350418/56735150-18230800-677e-11e9-8687-6802fdf81576.png)
![mobile-who-should-enroll](https://user-images.githubusercontent.com/45350418/56735163-253ff700-677e-11e9-8ec0-b3fd96fad099.png)

asadiqbal08 commented on cms/models.py:
```diff
@@ -12,6 +13,65 @@
+        help_text='Contents of the "Who Should Enroll" setion.',
+    )
+
+    content_panels = [StreamFieldPanel("content"), FieldPanel("image")]
```
nit: If any thing need to be add in future then will not mark all code changed. 

```
content_panels =  [
     StreamFieldPanel("content"),  
     FieldPanel("image"), 
]
```

ahmed-arbisoft replied:
Good catch

ahmed-arbisoft replied:
The formatter being used (black) isn't allowing this change as per configuration so I guess we'll have to stick with this for right now.

asadiqbal08 commented on cms/models.py:
```diff
@@ -79,6 +139,8 @@ class Meta:
         StreamFieldPanel("content"),
     ]
 
+    subpage_types = [WhoShouldEnrollPage]
```
is it not expecting a string in it ? Actually, I am passing string. 
`subpage_types = ['WhoShouldEnrollPage']`

ahmed-arbisoft replied:
You could pass in a string to lazy-load, that also works but we already have access to the class so it isn't really required to pass in a string, we can give it the actual model.

asadiqbal08 commented on cms/models.py:
```diff
@@ -12,6 +13,65 @@
+        # autogenerate a unique slug so we don't hit a ValidationError
+        self.title = "Who Should Enroll"
+        self.slug = slugify("{}-{}".format(self.get_parent().id, self.title))
+        super().save()
```
did you check it ? It is throwing a validation error on my end. I think we should keep the `title` that would be the title of page.

ahmed-arbisoft replied:
Do you mean this actual implementation, when checked out on your system, is throwing a `ValidationError`?

asadiqbal08 replied:
do not forget to pass  *args, **kwargs to parent. 
`super(WhoShouldEnrollPage, self).save(*args, **kwargs)`

ahmed-arbisoft replied:
Done.

asadiqbal08 commented on courses/models.py:
```diff
@@ -125,6 +126,15 @@ def duration(self):
+            if self.page and self.page.get_children().type(WhoShouldEnrollPage).count()
+            else None
+        )
+
```
Can we do it simple ? 

```
 """Gets the leGets the WhoShouldEnroll associated child page from the associate Page if it exists"""
        if self.page:
            result = self.page.get_children().type(WhoShouldEnrollPage).first()
            if result:
                return result.specific

        return None
```

ahmed-arbisoft replied:
I guess that's better.

ahmed-arbisoft replied:
Done.

asadiqbal08 commented on cms/models.py:
```diff
@@ -12,6 +13,65 @@
+    content_panels = [StreamFieldPanel("content"), FieldPanel("image")]
+
+    # disable promote panels, no need for slug entry, it will be autogenerated
+    promote_panels = []
```
should we not move it to parent ? I am thinking we will keep it disable for others too.

ahmed-arbisoft replied:
Done.

asadiqbal08 commented on courses/models_test.py:
```diff
@@ -1,12 +1,17 @@
+    ProgramPageFactory,
+    CoursePageFactory,
+    WhoShouldEnrollPageFactory,
+)
```
I am moving the test* files under `/courses/tests/` in my [PR](https://github.com/mitodl/mitxpro/pull/170)

ahmed-arbisoft replied:
Okay, I'll update my files once #170 goes through.

asadiqbal08 commented on courses/models_test.py:
```diff
@@ -244,3 +250,13 @@ def test_course_page_properties():
     assert course.video_title == "<p>title</p>"
     assert course.video_url == "http://test.com/mock.mp4"
     assert course.background_image.title == "background-image"
+    who_should_enroll_page = WhoShouldEnrollPageFactory(
```
Do we not need to add tests for ProgramPage properties ?

ahmed-arbisoft replied:
Done.

asadiqbal08 commented on cms/models.py:
```diff
@@ -72,6 +72,43 @@ def save(self, *args, **kwargs):
+
+    def save(self, *args, **kwargs):
+        # autogenerate a unique slug so we don't hit a ValidationError
+        self.title = "Who Should Enroll"
```
FYI. In my previous PR's, I used title like that. 
How you will learn -- Learning techniques.
What you will learn -- Learning outcomes.

asadiqbal08 commented on cms/models.py:
```diff
@@ -72,6 +72,43 @@ def save(self, *args, **kwargs):
+        # autogenerate a unique slug so we don't hit a ValidationError
+        self.title = "Who Should Enroll"
+        self.slug = slugify("{}-{}".format(self.get_parent().id, self.title))
+        super().save(*args, **kwargs)
```
I prefer: super(WhoShouldEnrollPage, self).save(*args, **kwargs)

asadiqbal08 replied:
Actually, I have used that pattern in #142 , so to keep code consistent that's why I was requesting.

asadiqbal08 commented on static/scss/detail/who-should-enroll.scss:
```diff
@@ -0,0 +1,93 @@
+
+    li {
+      font-size: 20px;
+      line-height: 28px;
```
in my [PR](https://github.com/mitodl/mitxpro/pull/200/files#diff-63410e023a547eafb5f22778c805b3d0)

ahmed-arbisoft replied:
Removed common values that were already available.

asadiqbal08 commented on courses/templates/course_detail.html:
```diff
@@ -95,4 +94,27 @@ <h3>{{ course.outcomes.sub_heading | safe }}</h3>
+      <div class="container">
+        <div class="row">
+          <div class="col-lg-6">
+            <h1>{{ course.who_should_enroll.title }}</h1>
```
I think, we can keep it hard coded. `WHO SHOULD ENROLL` here.

ahmed-arbisoft replied:
Done.

gsidebo commented:
@ahmed-arbisoft When I click "Add a Child Page" in Wagtail for a course page, I am taken directly to the "who should enroll" form. How do I add a learning outcomes page now? (#170)

ahmed-arbisoft commented:
If the page already has a learning outcomes page and you're left with only
one type of possible child page you can add, wagtail will take you directly
to the add form of that page type. If the page does not have a child page
and you try to add one, it will give you the choice to pick from the
possible candidates.

On Thu, 2 May 2019, 8:46 pm Gavin Sidebottom, <notifications@github.com>
wrote:

> @ahmed-arbisoft <https://github.com/ahmed-arbisoft> When I click "Add a
> Child Page" in Wagtail for a course page, I am taken directly to the "who
> should enroll" form. How do I add a learning outcomes page now? (#170
> <https://github.com/mitodl/mitxpro/pull/170>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitodl/mitxpro/pull/179#issuecomment-488725486>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKZ74EW7NNHUWOU2BMQX5XDPTMEDXANCNFSM4HIFLVXQ>
> .
>

-- 


The content of this email and attachment(s) are confidential and intended 
solely for the recipient(s) specified in the message. If you received this 
email by error, please alert the sender by replying to this email and 
follow with its deletion, so that we can ensure such an error does not 
occur in future. If you are not the intended recipient, you are strictly 
prohibited to use, copy, disseminate or store this email or its 
attachment(s).

gsidebo commented on static/scss/detail/who-should-enroll.scss:
```diff
@@ -0,0 +1,93 @@
+  h1 {
+    font-size: 40px;
+    line-height: 45px;
+    color: $who-should-enroll--heading;
```
I'm thinking we should do the following:
- Move this variable into this file (since it's only being used here), call it `$dark-bg-heading-color`, and define it at the top of the file
- Make a variable out of the block's bg color (`#2c2a29`) and call it `$darker-grey-bg`

ahmed-arbisoft replied:
Done

ahmed-arbisoft replied:
Done.

ahmed-arbisoft commented on courses/models_test.py:
```diff
@@ -112,8 +125,29 @@ def test_program_page_properties():
+    assert program.outcomes == learning_outcomes_page
     assert not LearningOutcomesPage.can_create_at(program_page)
 
+    # WhoShouldEnrollPage tests
```
Done

ahmed-arbisoft replied:
Done.

ahmed-arbisoft commented on static/scss/detail/who-should-enroll.scss:
```diff
@@ -0,0 +1,93 @@
+// sass-lint:disable mixins-before-declarations
+.who-should-enroll-block {
+  width: 100%;
```
Removed

ahmed-arbisoft replied:
Removed.

ahmed-arbisoft commented on courses/models_test.py:
```diff
@@ -286,3 +331,23 @@ def test_course_page_properties():
     assert course.outcomes == learning_outcomes_page
     assert not LearningOutcomesPage.can_create_at(course_page)
+
+    # WhoShouldEnrollPage tests
```
Done.

gsidebo commented:
@ahmed-arbisoft Did you see my last comment?

ahmed-arbisoft commented:
> @ahmed-arbisoft Did you see my last comment?

Done.

ahmed-arbisoft commented:
> I'm not seeing the fix. This is at screen width ~880px:
> 
> ![ss 2019-05-06 at 15 18 22 ](https://user-images.githubusercontent.com/14932219/57249202-53d98f80-7012-11e9-8144-c2fe9c10c703.png)
> 
> List items are still centered. I might be wrong but it seems to me that the list items should always be left-aligned no matter the screen width, and the section title should be centered at narrow widths

Hi, sorry I misread your comment and only moved the heading to be left aligned. I've fixed the alignment, left on list and center on title. Also added a layout switch (image on left, content on right) that was added by Peter.

tasawernawaz commented on courses/models_test.py:
```diff
@@ -172,6 +202,26 @@ def test_program_learning_techniques():
         assert technique.value.get("image").title == "image-title"
 
 
+def test_program_target_audience():
```
this test is pretty similar to `test_program_who_should_enroll` has same docstring as well, what is the real difference between them ?

ahmed-arbisoft replied:
Removed, `test_program_target_audience` is redundant.

tasawernawaz commented on courses/models.py:
```diff
@@ -177,6 +178,15 @@ def faqs(self):
 
+    @property
+    def who_should_enroll(self):
+        """Gets the WhoShouldEnroll associated child page from the associate Page if it exists"""
```
nit: associate --> associated

ahmed-arbisoft replied:
Fixed.

tasawernawaz commented on courses/templates/course_detail.html:
```diff
@@ -101,6 +101,29 @@ <h2>{{ course.outcomes.sub_heading | safe }}</h2>
+        <div class="container">
+          <div class="row">
+            <div class="col-lg-6 {% if course.who_should_enroll.switch_layout %}order-2{% endif %}">
+              <h1 class="text-uppercase">Who Should Enroll</h1>
```
currently we have a class `uppercase` in `static/scss/detail/common.scss`, It would be great if you please remove that and use this `text-uppercase` from bootstrap.

ahmed-arbisoft replied:
Removed from common, `_faqs.html` updated that was using it

tasawernawaz commented on courses/templates/_faqs.html:
```diff
@@ -30,7 +30,7 @@ <h3>Please review the following frequently asked questions before enrolling. Don
     <!--TODO related to above todo, we will load more faqs on button click-->
     {% if course.faqs|length > 5 %}
-    <a href="#" class="readmore uppercase"><span class="readmore-text">Read</span> More FAQs</a>
+    <a href="#" class="readmore text-uppercase"><span class="readmore-text">Read</span> More FAQs</a>
```
@ahmed-arbisoft please update on https://github.com/mitodl/mitxpro/blob/e64bc00a5ee1d869ada930e55118071fa2f5811f/courses/templates/_faqs.html#L13 as well.

ahmed-arbisoft replied:
Done.
